clash-lib/src/Clash/Normalize.hs

```diff
@@ -108,54 +114,75 @@ runNormalization
   -- ^ NormalizeSession to run
   -> IO a
-runNormalization env supply globals typeTrans peEval eval rcsMap topEnts =
-  runRewriteSession rwEnv rwState
-  where
-    -- TODO The RewriteEnv should just take ClashOpts.
-    rwEnv     = RewriteEnv
-                  env
-                  typeTrans
-                  peEval
-                  eval
-                  (mkVarSet topEnts)
-
-    rwState   = RewriteState
-                  0
-                  mempty       -- transformCounters Map
-                  globals
-                  supply
-                  (error $ $(curLoc) ++ "Report as bug: no curFun",noSrcSpan)
-                  0
-                  (IntMap.empty, 0)
-                  emptyVarEnv
-                  normState
-
-    normState = NormalizeState
-                  emptyVarEnv
-                  Map.empty
-                  emptyVarEnv
-                  emptyVarEnv
-                  Map.empty
-                  rcsMap
-
-normalize
-  :: [Id]
-  -> NormalizeSession BindingMap
-normalize []  = return emptyVarEnv
-normalize top = do
-  (new,topNormalized) <- unzip <$> mapM normalize' top
-  newNormalized <- normalize (concat new)
-  return (unionVarEnv (mkVarEnv topNormalized) newNormalized)
-
-normalize' :: Id -> NormalizeSession ([Id], (Id, Binding Term))
-normalize' nm = do
-  exprM <- lookupVarEnv nm <$> Lens.use bindings
+runNormalization env supply globals typeTrans peEval eval rcsMap lock entities session = do
+  normState <- NormalizeState
+    <$> MVar.newMVar emptyVarEnv
+    <*> MVar.newMVar Map.empty
+    <*> MVar.newMVar emptyVarEnv
+    <*> MVar.newMVar emptyVarEnv
+    <*> MVar.newMVar Map.empty
+    <*> MVar.newMVar rcsMap
+
+  rwState <- RewriteState
+    <$> MVar.newMVar mempty
+    <*> MVar.newMVar globals
+    <*> MVar.newMVar supply
+    <*> MVar.newMVar HashMap.empty
+    <*> MVar.newMVar 0
+    <*> MVar.newMVar (mempty, 0)
+    <*> MVar.newMVar emptyVarEnv
+    <*> pure lock
+    <*> pure normState
+
+  runRewriteSession rwEnv rwState session
+ where
+  rwEnv = RewriteEnv
+    { _clashEnv = env
+    , _typeTranslator = typeTrans
+    , _peEvaluator = peEval
+    , _evaluator = eval
+    , _topEntities = mkVarSet entities
+    }
+
+normalize :: [Id] -> NormalizeSession BindingMap
+normalize tops = do
+  q <- Monad.liftIO MS.newQ
+  traverse_ (Monad.liftIO . MS.pushL q) tops
+  binds <- MVar.newMVar (emptyVarSet, [])
+  -- one thread per top-level binding
+  Async.replicateConcurrently_ (length tops) (normalizeStep q binds)
+  mkVarEnv . snd <$> MVar.readMVar binds
+
+normalizeStep
+    :: MS.LinkedQueue Id
+    -> MVar (VarSet, [(Id, Binding Term)])
+    -> NormalizeSession ()
+normalizeStep q binds = do
+  res <- Monad.liftIO $ MS.tryPopR q
+  case res of
+      Just id' -> do
+        (bound, _) <- MVar.readMVar binds
+        unless (id' `elemVarSet` bound) $ do
+          -- immediately mark this work as being done
+          MVar.modifyMVar_ binds (pure . first (`extendVarSet` id'))
```

Comment:
You need to hold on to the binds-lock until line 174, or you risk race conditions, right?

Reply:
I'm also a bit suspicious of immediately marking it as done, isn't this misleading?

Reply:
> I think if you don't immediately mark it as "done" then two threads might work on it at once.

But I don't think you do it immediately, you need to

1. Get the lock for `binds`
2. Get the value of `binds`
3. Check for the id
4. Mark it as done
5. Release the lock for `binds`

Right now, I think this is happening:

1. Get the lock for `binds`
2. Get the value of `binds`
3. Release the lock for `binds`
4. Check for the id
5. Get the lock for `binds`
6. Mark it as done
7. Release the lock for `binds`

Because you acquire and release the lock twice, a competing thread can get hold of the value of `binds` and incorrectly conclude that it should _not_ skip.

The second part of my complaint is probably just a naming issue, like @alex-mckenna indicates :)

Reply:
Oh wow you're completely right, thank you for noticing! 

I've added a comment as well.